### PR TITLE
Change raw pointers to reference_wrapper types

### PR DESCRIPTION
**Pull request by davepl**
This checkin incorporates the previous branch again, I believe!

### CODE REVIEW

**Comment by rbergen** [comment]
> This checkin incorporates the previous branch again, I believe!

Almost. You didn't pull in the changes I made to fix the CI build issue on Ubuntu before you branched off `thursday`. But I see that later you came to the same change I had applied, so at the code level this branch is now equal on that front.

I do notice we're hop-skipping from one PR to the next without merging, and that means we don't really update `main` that often. I'm not saying that's a problem per se; I'm mentioning it so you can develop an opinion about it if you feel that's necessary. :)

File: controller.h [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+    vector<reference_wrapper<ICanvas>> Canvases() const override
+    {
+        vector<std::reference_wrapper<ICanvas>> canvases;
+        canvases.reserve(_canvases.size());
+        for (const auto& canvas : _canvases)
+            canvases.push_back(*canvas);
+        return canvases;
+    }
+
+    static std::unique_ptr<Controller> CreateFromFile(const std::string& filePath)
```
Comment (rbergen) [inline controller.h:54]: ...and that effectively kills the idea I had to easily override the standard set of canvases to match my one-Mesmerizer setup. I'll go back to the drawing board on that one. :)

File: serialization.h [hunk 2]
```diff
@@ -242,3 +244,76 @@ inline void from_json(const nlohmann::json& j, unique_ptr<ISocketChannel>& socke
...
+    }
+    catch (const std::exception &e)
+    {
+        j = nullptr;
+    }
+}
+
+// I've run into the same dependency loop issue again here... I'm not sure how to resolve it!
+// I think I need to move the from_json function to a separate file, but I know you want to keep
+// it here likely.  It can go right into Controller.h nicely though....
```
Comment (rbergen) [inline serialization.h:269]: No, I actually don't want to keep it here. I'd like us to make a decision on where to keep (de)serialization logic for types we declare ourselves, and stick with it.

Personally, I think we should keep the (de)serialization logic for a class/struct as close to the type's declaration as possible. This is also consistent with the approach we used in NDS, where the logic is actually _embedded_ in the classes/structs in question. 

In that scenario, serialization.h would only contain (de)serialization logic for a few (basic) types that are declared externally - again the same way we do things in NDS.

I'm happy to push an update to the Thursday branch to implement this refactor throughout the project, if you're okay with that. It would mean I remove the interface.h and serialization.h includes from monitor/main.cpp and monitor/content.cpp; we already concluded including those headers there is moot anyway.

File: webserver.h
Comment (rbergen) [inline webserver.h:1]: As part of the refactor I mentioned in my previous comment, I'd also consistently apply implicit (de)serialization throughout the code.

**Review by rbergen — COMMENTED** [review]
I have a few comments/suggestions, including an offer to refactor the distribution of (de)serialization code throughout the codebase.

While working on that, I'm also happy to fix the CI build errors we're now getting, due to the compiler being set to pedantic mode (-Wextra and -Werror) when compiling ledmon.

Let me know what you think!

File: Makefile [hunk 3]
```diff
@@ -29,8 +29,8 @@ endef
 # Detect platform
 UNAME_S := $(shell uname -s)
 ifeq ($(UNAME_S), Darwin)
-    CFLAGS += -I$(shell brew --prefix)/include/
-    LDFLAGS += -L$(shell brew --prefix)/lib/
+    CFLAGS += -I$(shell brew --prefix)/include/ -I$(shell brew --prefix spdlog)/include
+    LDFLAGS += -L$(shell brew --prefix)/lib/ -L$(shell brew --prefix fmt)/lib
```
[the remark this answers is not in the document]
Comment (davepl) [reply Makefile:33]: I do not understand those big words!  Feel free to change, but it works for me.  Not sure what keg only means!

File: serialization.h (hunk 2 above)
Comment (davepl) [reply serialization.h:269]: Yes, please do.  What you're describing is what I'd originally planned, but then tried to make it a priority to keep it all together, which didn't work out!

Comment (rbergen) [reply serialization.h:269]: Done. PR #16 aims to modify this PR to do exactly that. Which means that we can proceed with this PR when that one is merged.

File: effects/paletteeffect.h [hunk 4]
```diff
@@ -101,11 +100,12 @@ class PaletteEffect : public LEDEffectBase
...
     {
-        j = 
-        {
+        to_json(j["palette"], _Palette);  // Use existing palette serializer
+        
+        j = {
             {"type", "PaletteEffect"},
             {"name", Name()},
-            {"palette", _Palette}, // Assuming Palette has a ToJson method or serializer
+            {"palette", j["palette"]},
```
Comment (rbergen) [inline effects/paletteeffect.h:108]: I'm not sure I get this change. Isn't the previous approach more straightforward?

**Review by rbergen — COMMENTED** [review]
So I just have a couple of questions about the effect (de)serialization. One is in a comment below, the other is that you seem to be using your own FromJson/ToJson class-member function construct instead of nlohmann::json's from_json/to_json one, where the latter comes with all kind of (semi-)autoconversion magic.

I was wondering why that is?

**Comment by davepl** [comment]
The Palette serialization falls into the “written but never tested” code, I haven’t been paying too much attention to it. Effect serialization is next, and that’s where FromJson/ToJson come into play.  But I need Palette for ColorWaveEffect, so I started there.

Maybe they can be avoided but on deserialization, I didn’t think nlohlman::json would be able to do dynamic types.  Or maybe it can if I define a to_json and from_json for every effect type (I’ll start with a single effect and see if it can).  I might still need a case/if statement, but perhaps it can be written with the from_json/to_json, not sure.

Basically, I figure SOMEONE has to look at the type field at runtime, json won’t be able to do it dynamically.  Hence the ToJson/FromJson unless I can avoid it.

Any suggestions?

 -Dave

**Comment by davepl** [comment]
Just to confirm, I picked up main (83 changes!) and it compiles and runs!  Thanks for all the work on the serialization headers and so on!

Cheers,
Dave

**Comment by rbergen** [comment]
> Just to confirm, I picked up main (83 changes!) and it compiles and runs! Thanks for all the work on the serialization headers and so on! Cheers, Dave

Ah, great, thanks for confirming!
And you're welcome; you know I secretly enjoy this. ;)

**Comment by rbergen** [comment]
> [...] Maybe they can be avoided but on deserialization, I didn’t think nlohlman::json would be able to do dynamic types. Or maybe it can if I define a to_json and from_json for every effect type (I’ll start with a single effect and see if it can). I might still need a case/if statement, but perhaps it can be written with the from_json/to_json, not sure. Basically, I figure SOMEONE has to look at the type field at runtime, json won’t be able to do it dynamically. Hence the ToJson/FromJson unless I can avoid it. Any suggestions? -Dave

The problem with effect deserialization is that you have to deduce from JSON what effect type to deserialize to, which means mapping something in the JSON for an effect to that effect instance's type in C++. In NDS that's the reason for the whole list of EFFECT_* defines, ADD_EFFECT* macros and supporting types (EffectFactories mainly); they basically map a numeric value from JSON to an effect type to instantiate (additionally, they provide constructor parameters for the default set of effects, but we can ignore that for now). 

I'm not saying that is "the" way to do it, but the JSON-value-to-C++-type-mapping is something we'll have to solve one way or another.
